README.md

```diff
@@ -0,0 +1,536 @@
+Instead, all participating entities are clients connecting to a so-called _broker_.
+Clients _subscribe_ to _topics_, _publish_ messages to them, and the broker does the routing (and many other things).
+
+A good production-ready broker, like [EMQX](https://github.com/emqx/emqx) generally provides
+not only MQTT routing capabilities but many other intellectual features, like
+* other kinds of connection methods, like WebSockets;
+* different models of authentication and authorization;
+* streaming data to databases;
+* custom routing rules based on message introspection;
+* and so on.
+
+## Sensor Setup
+
+For the sake of simplicity, our device will be represented by an ordinary Mix
+application: it can be easily converted to a Nerves application.
+
+First, we create a Mix project:
+
+```bash
+mix new --sup weather_sensor
+cd weather_sensor
+```
+
+To interact with an MQTT broker, we need an MQTT client. We take [emqtt](https://github.com/emqx/emqtt).
+Add it to `mix.exs` as a dependency:
+
+```elixir
+defp deps do
+  [
+    {:emqtt, github: "emqx/emqtt", tag: "1.4.4"}
+  ]
+end
+```
+
+```bash
+# we do not want to use and build QUIC capabilities
+export BUILD_WITHOUT_QUIC=1
+mix deps.get
+```
+
+We will put all our "sensor" code to the main module `WeatherSensor`, so we need to
+add it to the application supervisor `lib/weather_sensor/application.ex`:
+
+```elixir
+defmodule WeatherSensor.Application do
+  @moduledoc false
+
+  use Application
+
+  @impl true
+  def start(_type, _args) do
+    children = [
+      WeatherSensor
+    ]
+
+    opts = [strategy: :one_for_one, name: WeatherSensor.Supervisor]
+    Supervisor.start_link(children, opts)
+  end
+end
+```
+
+Now, let's implement the main module in `lib/weather_sensor.ex`:
+
+```elixir
+defmodule WeatherSensor do
+  @moduledoc false
+
+  use GenServer
+
+  def start_link([]) do
+    GenServer.start_link(__MODULE__, [])
+  end
+
+  def init([]) do
+    interval = Application.get_env(:weather_sensor, :interval)
+    emqtt_opts = Application.get_env(:weather_sensor, :emqtt)
+    # This is the topic to which we will publish our reports
+    report_topic = "/reports/#{emqtt_opts[:clientid]}/temperature"
+    {:ok, pid} = :emqtt.start_link(emqtt_opts)
+
+    timer = :erlang.start_timer(interval, self(), :tick)
+    {:ok,
+      %{
+        interval: interval,
+        timer: timer,
+        report_topic: report_topic,
+        pid: pid
+      },
+      {:continue, :start_emqtt}}
+  end
+
+  def handle_continue(:start_emqtt, %{pid: pid} = st) do
+    # Here we make actual connection to the broker
+    {:ok, _} = :emqtt.connect(pid)
+    emqtt_opts = Application.get_env(:weather_sensor, :emqtt)
+    clientid = emqtt_opts[:clientid]
+    # From this topic we will receive commands
+    {:ok, _, _} = :emqtt.subscribe(pid, {"/commands/#{clientid}/set_interval", 1})
+    {:noreply, st}
+  end
+
+  def handle_info({:timeout, _ref, :tick}, %{report_topic: topic, pid: pid} = st) do
```

Comment:
Nit: was the `_ref` intended to match the `:timer` in the state?

Though, if `Process.send_after` would be used instead of `:erlang.start_timer`, the sent message won't contain the reference.